### PR TITLE
Use wrapped error that includes path

### LINKED ISSUE #3: Include path information in items
Walkdir includes the entry's path in a custom error type, because otherwise we just can't really understand the errors, since it's this crate that is issueing the function calls. Possibilities I have thought of so far:

1. Wrap the `std::io::Error` in a custom struct, like `Walkdir::Error`;
2. Make the `Stream`'s `Item` be a struct/tuple containing the path and the `Result`.

I'd be willing to work on this.

### PR DESCRIPTION
**Pull request by jdanford**
Closes #3

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "async-walkdir"
-version = "1.0.0"
+version = "1.1.0"
```
Comment (ririsoft) [inline Cargo.toml]: I believe the change requires a major version bump: we are replacing the previous error type `std::io::Error` by another one which is not binary compatible. (see https://doc.rust-lang.org/cargo/reference/semver.html#item-remove)

File: src/error.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+use std::{
```
Comment (ririsoft) [inline src/error.rs:5]: Can you had a copyright & license header please ?
For instance :

```rust
// SPDX-FileCopyrightText: 2020-2024 Ririsoft <riri@ririsoft.com>
// SPDX-FileCopyrightText: 2024 Your whatever name/nickname <your whatever mail>
// SPDX-License-Identifier: Apache-2.0
```

File: src/error.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    error, fmt, io, ops::Deref, path::{Path, PathBuf}
+};
+
+/// A wrapper around [`io::Error`] that includes the associated path.
+#[derive(Debug)]
```
Comment (ririsoft) [inline src/error.rs]: What about using [thiserror](https://docs.rs/thiserror/latest/thiserror/) ?
We could have something like :

```rust
#[derive(Error, Debug)]
pub enum Error {
    #[error("IO error at '{path}': {}")]
    Io {
        path: PathBuf,
        #[from]
        source: io::Error,
    },
}
```

PROs: We are massively simplifying the code.
CONs: We are adding a dependency on `thiserror`, not big deal to me as it is a quiet standard pattern.

**Review by ririsoft — CHANGES_REQUESTED** [review]
Hello @jdanford ,

Thank you very much for your proposal, this will allow to close a long standing issue.
I added a few comments to the code review, with some required changes and open questions where your feedback would be appreciated.

Thank you very much in advance !

**Comment by ririsoft** [comment]
Also can you format the code with `cargo fmt` so that the CI/CD runs OK ?

**Comment by jdanford** [comment]
Alright, made all those changes. Using `thiserror` sounds fine, it's your project so I'll let you make that call.

**Comment by ririsoft** [comment]
Hello @jdanford, I pushed to your branch a new version using [`thiserror` ](https://docs.rs/thiserror/latest/thiserror/index.html). Does it matches your expectations ?

**Comment by jdanford** [comment]
Works for me!

**Review by ririsoft — APPROVED** [review]

**Comment by ririsoft** [comment]
Thank you @jdanford for your contribution. I will release soon, once I also close #4.

**Pull request merged**